### PR TITLE
Improve peer list peer availability error messages

### PR DESCRIPTION
**Pull request by kriskowal**

### CODE REVIEW

File: compressor/gzip/gzip.go [hunk 1]
```diff
@@ -18,32 +18,126 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-// Package gzip provides a YARPC binding for gzip compression.
-package gzip
+// Package yarpcgzip provides a YARPC binding for gzip compression.
+package yarpcgzip
```
Comment (chucknthem) [inline compressor/gzip/gzip.go]: this change looks unrelated...

Comment (kriskowal) [reply compressor/gzip/gzip.go]: Thanks. The perils of dropping everything to fix a thing…

File: peer/abstractlist/list.go [hunk 2]
```diff
@@ -548,31 +557,38 @@ func (pl *List) newNoContextDeadlineError() error {
 }
 
 func (pl *List) newUnavailableError(err error) error {
-	return yarpcerrors.Newf(yarpcerrors.CodeUnavailable, "%q peer list timed out waiting for peer: %s", pl.name, err.Error())
+	return yarpcerrors.Newf(yarpcerrors.CodeUnavailable, "%q peer list timed out waiting for peer, %s assigned, fail-fast is not enabled: %s", pl.name, pl.formatNumPeers(), err.Error())
```
Comment (AlexAPB) [inline peer/abstractlist/list.go]: - Could we indicate the availability of peer we're waiting for? ie `... timed out waiting for an available peer`
- and instead of `assigned` what if we used `unavailable`

This would read like:
```
"%q peer list timed out waiting for an available peer (%s unavailable), fail-fast is disabled: %s"
```

Comment (prashantv) [inline peer/abstractlist/list.go]: little nits:
 - The formatting of this message is inconsistent with the "fail-fast is enabled" message. This message has # assigned after fail-fast status, while the other message has fail-fast status before # assigned
 - While I can guess what "assigned" peers means in this context, I'm not sure it's user-friendly
 - I still think separating out the 0 peers case is more useful than mixing the messages and just having the number be 0 or 1

Looking at the stakeholder experience, I think the scenarios/error groupings are
 - No peers in the peer list. We waited for a peer to show up but timed out (fail-fast is not enabled)
 - No peers in the peer list, and we didn't wait (fail-fast enabled)
 - There are X peers in the peer-list, but none of them are usable, and we timed out waiting. We need to also find a good word for "usable" and help the user understand why they may not be usable
 - Same as above, but we didn't wait (fail-fast enabled).

Comment (kriskowal) [reply peer/abstractlist/list.go]: We could just say “connected” or possibly “responsive”.

**Comment by kriskowal** [comment]
Added a fixup commit to the top of the stack. This produces error messages like:

- has no peers,
  - did not wait for peers to be added (fail-fast is enabled)
  - waited for peers to be added but timed out (fail-fast is not enabled):
    - context cancelled
    - context deadline exceeded
- has 1 peer but it is unresponsive,
- has 10 peers but none are responsive,
  - did not wait for a peer to respond (fail-fast is enabled)
  - timed out waiting for a peer to respond (fail-fast is not enabled):
    - context cancelled
    - context deadline exceeded

**Comment by kriskowal** [comment]
Changed "did not wait for a peer to respond" to "did not wait for a connection to open (fail-fast is enabled)", &c.
